### PR TITLE
Add peers bootstrap merge local writes integration test

### PR DESCRIPTION
**Pull request by robskillington**
cc @xichen2020 @martin-mao @cw9

### CODE REVIEW

File: integration/data.go [hunk 1]
```diff
@@ -97,36 +99,88 @@ func verifySeriesMapForRange(
...
+	)
+	if debugFilePathPrefix != "" {
+		for _, entry := range []struct {
+			suffix   string
+			filePath *string
+		}{
+			{"expected.log", &expectedDebugFilePath},
+			{"actual.log", &actualDebugFilePath},
+		} {
+			*entry.filePath = debugFilePathPrefix + "_" + entry.suffix
```
Comment (cw9) [inline integration/data.go]: nit: put the file operations in a function and call it on "expected.log" and "actual.log" separately? not sure if it worth to loop through 2 structs

Comment (robskillington) [inline integration/data.go]: This is just a test harness however?

Comment (robskillington) [inline integration/data.go]: Will call another function though, guess it's not too much of a pollution.

File: integration/data.go [hunk 2]
```diff
@@ -97,37 +99,84 @@ func verifySeriesMapForRange(
...
+	}
+	if len(actualDebugFilePath) > 0 {
+		writeVerifyDebugOutput(t, actualDebugFilePath, start, end, actual)
+	}
+
 	require.Equal(t, expected, actual)
 }
 
+func writeVerifyDebugOutput(t *testing.T, filePath string, start, end time.Time, series seriesList) {
+	w, err := os.OpenFile(filePath, os.O_APPEND|os.O_WRONLY, os.ModeAppend)
```
Comment (martin-mao) [inline integration/data.go]: Does this get cleaned up later since it's append?

File: storage/namespace.go [hunk 3]
```diff
@@ -216,18 +219,35 @@ func (n *dbNamespace) Bootstrap(
 	}
 	n.metrics.bootstrap.Success.Inc(1)
 
+	// Bootstrap shards using at least half the CPUs available
+	workers := pool.NewWorkerPool(int(math.Ceil(float64(runtime.NumCPU()) / 2)))
```
Comment (martin-mao) [inline storage/namespace.go]: Why not just do runtime.NumCPU() / 2 and leave it to the int rounding.

File: integration/data.go (hunk 2 above)
Comment (robskillington) [inline integration/data.go]: This is an integration test component that is only enabled "when you want it".  It's for viewing the difference between expected and actual datapoints that arrived at the destination node.  It needs to live on after the test run is finished so you can do a diff between the two files.

File: storage/namespace.go [hunk 4]
```diff
@@ -216,18 +219,35 @@ func (n *dbNamespace) Bootstrap(
...
+			var bootstrappedSeries map[string]block.DatabaseSeriesBlocks
+			if result, ok := results[shard.ID()]; ok {
+				bootstrappedSeries = result.AllSeries()
+			}
 
-		err := shard.Bootstrap(bootstrappedSeries, writeStart, cutover)
-		multiErr = multiErr.Add(err)
+			err := shard.Bootstrap(bootstrappedSeries, writeStart, cutover)
+
+			mutex.Lock()
```
Comment (martin-mao) [inline storage/namespace.go]: Nit: move unlock to line below multiErr to a void defer overhead.

File: storage/namespace.go (hunk 3 above)
Comment (robskillington) [inline storage/namespace.go]: I don't want it to round down to 0 if NumCPU() == 1.  e.g. https://play.golang.org/p/J8MuG1wGNF

File: storage/namespace.go (hunk 4 above)
Comment (robskillington) [inline storage/namespace.go]: Sure thing.

File: storage/series.go [hunk 5]
```diff
@@ -333,33 +338,54 @@ func (s *dbSeries) bufferDrained(start time.Time, encoder encoding.Encoder) {
...
-		}
-		blockStart := dp.Timestamp.Truncate(s.opts.RetentionOptions().BlockSize())
-		block := blocks.BlockOrAdd(blockStart)
+	multiIter := s.opts.MultiReaderIteratorPool().Get()
+	multiIter.Reset(readers)
+	defer multiIter.Close()
 
-		if err := block.Write(dp.Timestamp, dp.Value, unit, annotation); err != nil {
+	blopts := s.opts.DatabaseBlockOptions()
+	encoder := s.opts.EncoderPool().Get()
```
Comment (martin-mao) [inline storage/series.go]: Need to defer close the encoder for return on line 380?

Comment (robskillington) [inline storage/series.go]: Encoders get left open and it is up to the block to close it when the block is finished with.

File: storage/namespace.go [hunk 6]
```diff
@@ -216,18 +219,35 @@ func (n *dbNamespace) Bootstrap(
 	}
 	n.metrics.bootstrap.Success.Inc(1)
 
+	// Bootstrap shards using at least half the CPUs available
+	workers := pool.NewWorkerPool(int(math.Ceil(float64(runtime.NumCPU()) / 2)))
+	workers.Init()
+
 	multiErr := xerrors.NewMultiError()
+	mutex := sync.Mutex{}
```
Comment (xichen2020) [inline storage/namespace.go]: nit: using `var mutex sync.Mutex` is probably more idiomatic

**Comment by xichen2020** [comment]
other than the nit LGTM

File: storage/series.go (hunk 5 above)
Comment (martin-mao) [inline storage/series.go]: But it won't get to the block until line 382. What happens in the case where we get an error on line 380, the encoder would be fetched from the pool and never returned no?

**Pull request merged**